File: MIDTERM-LAB-3/Java/MP19.java
import java.io.*;
import java.util.*;

public class MP19 {
    public static void main(String[] args) {
        Scanner input = new Scanner(System.in);
        System.out.print("Enter dataset file path: ");
        String path = input.nextLine();

        File file = new File(path);
        if (!file.exists() || !file.isFile()) {
            System.out.println("Invalid file path.");
            return;
        }

        try (BufferedReader br = new BufferedReader(new FileReader(file))) {
            String header = br.readLine(); // ヘッダー行
            String line;
            int total = 0;
            int passCount = 0, failCount = 0;
            int sumScore = 0;
            int maxScore = Integer.MIN_VALUE;
            int minScore = Integer.MAX_VALUE;

            while ((line = br.readLine()) != null) {
                String[] values = line.split(",");
                if (values.length < 8) continue; // 不完全行はスキップ

                int score = Integer.parseInt(values[6]);
                String result = values[7];

                sumScore += score;
                maxScore = Math.max(maxScore, score);
                minScore = Math.min(minScore, score);
                total++;

                if (result.equals("PASS")) passCount++;
                else if (result.equals("FAIL")) failCount++;
            }

            System.out.println("=== Dataset Summary Report ===");
            System.out.println("Total records: " + total);
            System.out.println("Average score: " + (sumScore / (double) total));
            System.out.println("Highest score: " + maxScore);
            System.out.println("Lowest score: " + minScore);
            System.out.println("Pass count: " + passCount);
            System.out.println("Fail count: " + failCount);

        } catch (IOException e) {
            System.out.println("Error reading file: " + e.getMessage());
        }
    }
}

File: MIDTERM-LAB-3/JavaScript/MP19.js
const fs = require('fs');
const readline = require('readline');
const path = require('path');

function validateFilePath(filePath) {
    if (!fs.existsSync(filePath)) {
        console.log("Error: File does not exist.");
        return false;
    }
    if (path.extname(filePath) !== '.csv') {
        console.log("Error: File is not in CSV format.");
        return false;
    }
    return true;
}

function parseCSV(filePath) {
    const data = fs.readFileSync(filePath, 'utf8');
    const lines = data.split('\n').filter(line => line.trim() !== '');
    const headers = lines[0].split(',');
    const records = lines.slice(1).map(line => {
        const values = line.split(',');
        return Object.fromEntries(headers.map((h, i) => [h.trim(), values[i] ? values[i].trim() : null]));
    });
    return records;
}

function generateSummary(records) {
    let total = records.length;
    let sumScore = 0, passCount = 0, failCount = 0;
    let maxScore = -Infinity, minScore = Infinity;

    records.forEach(r => {
        const score = parseInt(r.Score);
        const result = r.Result;
        if (!isNaN(score)) {
            sumScore += score;
            maxScore = Math.max(maxScore, score);
            minScore = Math.min(minScore, score);
        }
        if (result === "PASS") passCount++;
        else if (result === "FAIL") failCount++;
    });

    console.log("=== Dataset Summary Report ===");
    console.log("Total records:", total);
    console.log("Average score:", (sumScore / total).toFixed(2));
    console.log("Highest score:", maxScore);
    console.log("Lowest score:", minScore);
    console.log("Pass count:", passCount);
    console.log("Fail count:", failCount);
}

async function main() {
    const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
    let filePath;
    while (true) {
        filePath = await new Promise(resolve => rl.question("Enter dataset file path: ", resolve));
        if (validateFilePath(filePath)) break;
    }
    const records = parseCSV(filePath);
    generateSummary(records);
    rl.close();
}

main();
